frame/vesting/src/lib.rs

```diff
@@ -312,105 +327,407 @@ pub mod pallet {
+		)?;
+
+		// We can't let this fail because the currency transfer has already happened
+		Self::add_vesting_schedule(
+			&target,
+			schedule.locked(),
+			schedule.per_block(),
+			schedule.starting_block(),
+		)
+		.expect("schedule inputs and vec bounds have been validated. q.e.d.");
+
+		Ok(())
+	}
+
+	/// Iterate through the schedules to track the current locked amount and
+	/// filter out completed and specified schedules.
+	///
+	/// Returns a tuple that consists of:
+	/// - bounded vec of vesting schedules, where completed schedules and those
+	/// specified by filter are remove.
+	/// - the amount locked at the current block number based on the given schedules.
+	///
+	/// NOTE: the amount locked does not include any schedules that are filtered out.
+	fn report_schedule_updates(
+		schedules: BoundedVec<VestingInfo<BalanceOf<T>, T::BlockNumber>, T::MaxVestingSchedules>,
+		action: VestingAction,
+	) -> (BoundedVec<VestingInfo<BalanceOf<T>, T::BlockNumber>, T::MaxVestingSchedules>, BalanceOf<T>)
+	{
 		let now = <frame_system::Pallet<T>>::block_number();
-		let locked_now = vesting.locked_at::<T::BlockNumberToBalance>(now);
 
-		if locked_now.is_zero() {
-			T::Currency::remove_lock(VESTING_ID, &who);
-			Vesting::<T>::remove(&who);
-			Self::deposit_event(Event::<T>::VestingCompleted(who));
+		let mut total_locked_now: BalanceOf<T> = Zero::zero();
+		let filtered_schedules = schedules
+			.into_iter()
+			.enumerate()
+			.filter_map(|(index, schedule)| {
+				let locked_now = schedule.locked_at::<T::BlockNumberToBalance>(now);
+				if locked_now.is_zero() || action.should_remove(&index) {
+					None
+				} else {
+					// We track the locked amount only if the schedule is included.
+					total_locked_now = total_locked_now.saturating_add(locked_now);
+					Some(schedule)
+				}
+			})
+			.collect::<Vec<_>>()
+			.try_into()
+			.expect("`BoundedVec` is created from another `BoundedVec` with same bound; q.e.d.");
+
+		(filtered_schedules, total_locked_now)
+	}
+
+	/// Write an accounts updated vesting lock to storage.
+	fn write_lock(who: &T::AccountId, total_locked_now: BalanceOf<T>) {
+		if total_locked_now.is_zero() {
+			T::Currency::remove_lock(VESTING_ID, who);
+			Self::deposit_event(Event::<T>::VestingCompleted(who.clone()));
 		} else {
 			let reasons = WithdrawReasons::TRANSFER | WithdrawReasons::RESERVE;
-			T::Currency::set_lock(VESTING_ID, &who, locked_now, reasons);
-			Self::deposit_event(Event::<T>::VestingUpdated(who, locked_now));
+			T::Currency::set_lock(VESTING_ID, who, total_locked_now, reasons);
+			Self::deposit_event(Event::<T>::VestingUpdated(who.clone(), total_locked_now));
+		};
+	}
+
+	/// Write an accounts updated vesting schedules to storage.
+	fn write_vesting(
+		who: &T::AccountId,
+		schedules: BoundedVec<VestingInfo<BalanceOf<T>, T::BlockNumber>, T::MaxVestingSchedules>,
+	) -> Result<(), DispatchError> {
+		if schedules.len() > T::MaxVestingSchedules::get() as usize {
```

Comment:
is this possible since the schedules are already bounded? probably you can simplify all the APIs by accepting a vec until here, and then bounding the vec in this call (or failing)

Reply:
40fb449